### PR TITLE
Improved version sorting

### PR DESCRIPTION
**Pull request by kaifox**
Please have a look ... sorting of versions...

The main work is done in the VersionStrings class....

### CODE REVIEW

File: src/java/org/jmad/modelpack/domain/ModelPackages.java [hunk 1]
```diff
@@ -8,24 +8,44 @@
...
  * 
  * @author kfuchsbe
  */
 public final class ModelPackages {
 
+    private static final Comparator<Variant> LATEST_RELEASE_LAST_COMPARATOR = (Variant v1, Variant v2) -> {
+        if (isRelease(v1) && isRelease(v2)) {
+            return VersionStrings.versionComparator().compare(v1.name(), v2.name());
+        }
+        return 0;
```
Comment (michi42) [inline src/java/org/jmad/modelpack/domain/ModelPackages.java]: Are you sure this fulfills the Comparator contract?
https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html#compare-T-T-

For sure, it is not consistent with equals() (that's fine but could be documented), but assume
X = Release 1
Y = Branch
Z = Release 3
so ``compare(X, Y) = 0``, ``compare(X, Z) = 1``, ``compare (Y, Z) = 0``
... which I think violates ``(compare(X, Y) == 0)`` must imply that ``sgn(compare(X, Z)) == sgn(compare(Y,Z))``

I think you at least need to add that ``compare(<Release>, <NO Release>)`` = 1 and ``compare(<NO Release>, <Release>)`` = -1 - or vice versa if you prefer.

File: src/java/org/jmad/modelpack/util/VersionStrings.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        if (v2digits.size() > smallestSize) {
+            return -1;
+        } else if (v1digits.size() > smallestSize) {
+            return 1;
+        }
+
+        return naturalStringOrder(v1, v2);
+    };
+
+    private static int naturalStringOrder(String v1, String v2) {
```
Comment (michi42) [inline src/java/org/jmad/modelpack/util/VersionStrings.java]: Why not just ``v1.compareTo(v2)`` ?

**Review by michi42 — COMMENTED** [review]
Looks good to me in general. I think the comparator to pick the latest release breaks the contract and should be changed (to avoid unpleasant surprises in the future ...)

File: src/java/org/jmad/modelpack/domain/ModelPackages.java (hunk 1 above)
Comment (kaifox) [reply src/java/org/jmad/modelpack/domain/ModelPackages.java]: Indeed.. I ignored this fact at the moment as the comparator is used in a chain, where the types are checked before ... However, probably you are right and one better fixes this here ... I will do so.

File: src/java/org/jmad/modelpack/util/VersionStrings.java (hunk 2 above)
Comment (kaifox) [reply src/java/org/jmad/modelpack/util/VersionStrings.java]: Indeed ;-)

**Review by michi42 — APPROVED** [review]
Looks good to me now, thanks for the fixes.

**Pull request merged**